# Data Wrangling

## Contents<a id='2_Contents'></a>
* [1. Introduction](#1_Introduction)
  * [1.1. Purpose of the Problem](#1.1_Purpose_of_the_Problem)
* [2.1. Importing necessary packages](#2.1_Importing_necessary_packages)
* [2.2. Load the dataset](#2.2_Load_the_dataset)
* [2.3. Explore the data](#2.3_Explore_the_data)

## 1. Introduction<a id='2.1_Introduction'></a>

In this step, data will be collected and organized, to make sure that we have clean, well-constructed and well-defined data. This step is crucial and prevents potential wrong results at the end.

### 1.1 Purpose of the Problem<a id= '2.1.1_Purpose_of_the_Problem'></a>

The importance and purpose of the project is to build a model which predicts which customers most likely to be retained.

## 2.1. Importing necessary packages<a id= '2.2_Importing_necessary_packages'></a>

In [1]:
# Import pandas, matplotlib.pyplot, seaborn, datetime, numpy-all necessary packages
import os
import pandas as pd
import matplotlib.pyplot as plt
import datetime
import numpy as np
import seaborn as sns

## 2.2. Load the dataset<a id= '2.3_Load_the_dataset'></a>

In [2]:
# Upload the dataset
from google.colab import files
uploades = files.upload()
df = pd.read_csv("IBM-Marketing-Customer-Value-Analysis.csv", encoding = "latin-1")

Saving IBM-Marketing-Customer-Value-Analysis.csv to IBM-Marketing-Customer-Value-Analysis.csv


## 2.3. Explore the data <a id='2.4_Explore_the_data'></a>

In [3]:
# Review the first few rows of the dataset
df.head()

Customer       State  Customer Lifetime Value Response  Coverage Education  \
0  BU79786  Washington              2763.519279       No     Basic  Bachelor   
1  QZ44356     Arizona              6979.535903       No  Extended  Bachelor   
2  AI49188      Nevada             12887.431650       No   Premium  Bachelor   
3  WW63253  California              7645.861827       No     Basic  Bachelor   
4  HB64268  Washington              2813.692575       No     Basic  Bachelor   

  Effective To Date EmploymentStatus Gender  Income  ...  \
0           2/24/11         Employed      F   56274  ...   
1           1/31/11       Unemployed      F       0  ...   
2           2/19/11         Employed      F   48767  ...   
3           1/20/11       Unemployed      M       0  ...   
4            2/3/11         Employed      M   43836  ...   

  Months Since Policy Inception Number of Open Complaints  Number of Policies  \
0                             5                         0                   1   
1                            42                         0                   8   
2                            38                         0                   2   
3                            65                         0                   7   
4                            44                         0                   1   

      Policy Type        Policy  Renew Offer Type  Sales Channel  \
0  Corporate Auto  Corporate L3            Offer1          Agent   
1   Personal Auto   Personal L3            Offer3          Agent   
2   Personal Auto   Personal L3            Offer1          Agent   
3  Corporate Auto  Corporate L2            Offer1    Call Center   
4   Personal Auto   Personal L1            Offer1          Agent   

  Total Claim Amount  Vehicle Class Vehicle Size  
0         384.811147   Two-Door Car      Medsize  
1        1131.464935  Four-Door Car      Medsize  
2         566.472247   Two-Door Car      Medsize  
3         529.881344            SUV      Medsize  
4         138.130879  Four-Door Car      Medsize  

[5 rows x 24 columns]

In [4]:
df.tail()

Customer       State  Customer Lifetime Value Response  Coverage  \
9129  LA72316  California             23405.987980       No     Basic   
9130  PK87824  California              3096.511217      Yes  Extended   
9131  TD14365  California              8163.890428       No  Extended   
9132  UP19263  California              7524.442436       No  Extended   
9133  Y167826  California              2611.836866       No  Extended   

     Education Effective To Date EmploymentStatus Gender  Income  ...  \
9129  Bachelor           2/10/11         Employed      M   71941  ...   
9130   College           2/12/11         Employed      F   21604  ...   
9131  Bachelor            2/6/11       Unemployed      M       0  ...   
9132   College            2/3/11         Employed      M   21941  ...   
9133   College           2/14/11       Unemployed      M       0  ...   

     Months Since Policy Inception Number of Open Complaints  \
9129                            89                         0   
9130                            28                         0   
9131                            37                         3   
9132                             3                         0   
9133                            90                         0   

      Number of Policies     Policy Type        Policy  Renew Offer Type  \
9129                   2   Personal Auto   Personal L1            Offer2   
9130                   1  Corporate Auto  Corporate L3            Offer1   
9131                   2  Corporate Auto  Corporate L2            Offer1   
9132                   3   Personal Auto   Personal L2            Offer3   
9133                   1  Corporate Auto  Corporate L3            Offer4   

      Sales Channel Total Claim Amount  Vehicle Class Vehicle Size  
9129            Web         198.234764  Four-Door Car      Medsize  
9130         Branch         379.200000  Four-Door Car      Medsize  
9131         Branch         790.784983  Four-Door Car      Medsize  
9132         Branch         691.200000  Four-Door Car        Large  
9133    Call Center         369.600000   Two-Door Car      Medsize  

[5 rows x 24 columns]

In [5]:
df.shape

(9134, 24)

In [6]:
# Review the data summary
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9134 entries, 0 to 9133
Data columns (total 24 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Customer                       9134 non-null   object 
 1   State                          9134 non-null   object 
 2   Customer Lifetime Value        9134 non-null   float64
 3   Response                       9134 non-null   object 
 4   Coverage                       9134 non-null   object 
 5   Education                      9134 non-null   object 
 6   Effective To Date              9134 non-null   object 
 7   EmploymentStatus               9134 non-null   object 
 8   Gender                         9134 non-null   object 
 9   Income                         9134 non-null   int64  
 10  Location Code                  9134 non-null   object 
 11  Marital Status                 9134 non-null   object 
 12  Monthly Premium Auto           9134 non-null   i

In [7]:
# Review the data types
df.dtypes

Customer                          object
State                             object
Customer Lifetime Value          float64
Response                          object
Coverage                          object
Education                         object
Effective To Date                 object
EmploymentStatus                  object
Gender                            object
Income                             int64
Location Code                     object
Marital Status                    object
Monthly Premium Auto               int64
Months Since Last Claim            int64
Months Since Policy Inception      int64
Number of Open Complaints          int64
Number of Policies                 int64
Policy Type                       object
Policy                            object
Renew Offer Type                  object
Sales Channel                     object
Total Claim Amount               float64
Vehicle Class                     object
Vehicle Size                      object
dtype: object

There are 9134 rows and 24 columns. The row starts at 0 and ends at 9133. 
There are numerical (all are continuous) and categorical data. 

Categorical data- Customer, State, Response, Coverage, Education, Effective To Date, EmploymentStatus, Gender, Location Code, Marital Status, Policy Type, Policy, Renew Offer Type, Sales Channel, Vehicle Class, Vehicle Size

Numerical/continuous data- Customer Lifetime value, Income, Monthly Premium Auto, Months Since Last Claim, Months Since Policy Inception, Number of Open Complaints, Number of Policies, Total Claim Amount.

###### Always make sure that numerical and categorucal variables are correctly classified. Not every numerical value represents non-categorical variables. Moreover, sometimes datetime data is being classified as 'object', in this case it is cruical to convert it to a datetime object.

In [8]:
# Identify number of non-null values
df.isna().sum().sum()

0

There are no non-null values, therefore no further action is required to replace the non-null values

In [9]:
# Find out percentage of the unique values per column
for name in df.columns:
   print(name + " - " + str(round(100 * (df[name].nunique()/df[name].size),2)) + "%")

Customer - 100.0%
State - 0.05%
Customer Lifetime Value - 88.03%
Response - 0.02%
Coverage - 0.03%
Education - 0.05%
Effective To Date - 0.65%
EmploymentStatus - 0.05%
Gender - 0.02%
Income - 62.34%
Location Code - 0.03%
Marital Status - 0.03%
Monthly Premium Auto - 2.21%
Months Since Last Claim - 0.39%
Months Since Policy Inception - 1.09%
Number of Open Complaints - 0.07%
Number of Policies - 0.1%
Policy Type - 0.03%
Policy - 0.1%
Renew Offer Type - 0.04%
Sales Channel - 0.04%
Total Claim Amount - 55.9%
Vehicle Class - 0.07%
Vehicle Size - 0.03%


In [10]:
# Number of the unique values of each column
df.nunique()

Customer                         9134
State                               5
Customer Lifetime Value          8041
Response                            2
Coverage                            3
Education                           5
Effective To Date                  59
EmploymentStatus                    5
Gender                              2
Income                           5694
Location Code                       3
Marital Status                      3
Monthly Premium Auto              202
Months Since Last Claim            36
Months Since Policy Inception     100
Number of Open Complaints           6
Number of Policies                  9
Policy Type                         3
Policy                              9
Renew Offer Type                    4
Sales Channel                       4
Total Claim Amount               5106
Vehicle Class                       6
Vehicle Size                        3
dtype: int64

In [11]:
# Identify the range of the columns
df.agg([min,max]).T

min           max
Customer                              AA10041       ZZ97035
State                                 Arizona    Washington
Customer Lifetime Value           1898.007675   83325.38119
Response                                   No           Yes
Coverage                                Basic       Premium
Education                            Bachelor        Master
Effective To Date                      1/1/11        2/9/11
EmploymentStatus                     Disabled    Unemployed
Gender                                      F             M
Income                                      0         99981
Location Code                           Rural         Urban
Marital Status                       Divorced        Single
Monthly Premium Auto                       61           298
Months Since Last Claim                     0            35
Months Since Policy Inception               0            99
Number of Open Complaints                   0             5
Number of Policies                          1             9
Policy Type                    Corporate Auto  Special Auto
Policy                           Corporate L1    Special L3
Renew Offer Type                       Offer1        Offer4
Sales Channel                           Agent           Web
Total Claim Amount                   0.099007   2893.239678
Vehicle Class                   Four-Door Car  Two-Door Car
Vehicle Size                            Large         Small

In [12]:
# Review the summary statistics
df.select_dtypes(include = 'number').describe().T

count          mean           std  \
Customer Lifetime Value        9134.0   8004.940475   6870.967608   
Income                         9134.0  37657.380009  30379.904734   
Monthly Premium Auto           9134.0     93.219291     34.407967   
Months Since Last Claim        9134.0     15.097000     10.073257   
Months Since Policy Inception  9134.0     48.064594     27.905991   
Number of Open Complaints      9134.0      0.384388      0.910384   
Number of Policies             9134.0      2.966170      2.390182   
Total Claim Amount             9134.0    434.088794    290.500092   

                                       min          25%           50%  \
Customer Lifetime Value        1898.007675  3994.251794   5780.182197   
Income                            0.000000     0.000000  33889.500000   
Monthly Premium Auto             61.000000    68.000000     83.000000   
Months Since Last Claim           0.000000     6.000000     14.000000   
Months Since Policy Inception     0.000000    24.000000     48.000000   
Number of Open Complaints         0.000000     0.000000      0.000000   
Number of Policies                1.000000     1.000000      2.000000   
Total Claim Amount                0.099007   272.258244    383.945434   

                                        75%           max  
Customer Lifetime Value         8962.167041  83325.381190  
Income                         62320.000000  99981.000000  
Monthly Premium Auto             109.000000    298.000000  
Months Since Last Claim           23.000000     35.000000  
Months Since Policy Inception     71.000000     99.000000  
Number of Open Complaints          0.000000      5.000000  
Number of Policies                 4.000000      9.000000  
Total Claim Amount               547.514839   2893.239678

In [13]:
# Identify number of missing values by column
na_values = pd.concat([df.isnull().sum(), 100 * df.isnull().mean()], axis=1)
na_values.columns=['Count', '%']
na_values.sort_values(by= ['Count'], ascending=False)

Count    %
Customer                           0  0.0
State                              0  0.0
Vehicle Class                      0  0.0
Total Claim Amount                 0  0.0
Sales Channel                      0  0.0
Renew Offer Type                   0  0.0
Policy                             0  0.0
Policy Type                        0  0.0
Number of Policies                 0  0.0
Number of Open Complaints          0  0.0
Months Since Policy Inception      0  0.0
Months Since Last Claim            0  0.0
Monthly Premium Auto               0  0.0
Marital Status                     0  0.0
Location Code                      0  0.0
Income                             0  0.0
Gender                             0  0.0
EmploymentStatus                   0  0.0
Effective To Date                  0  0.0
Education                          0  0.0
Coverage                           0  0.0
Response                           0  0.0
Customer Lifetime Value            0  0.0
Vehicle Size                       0  0.0

There are no missing values, therefore no further action is required to replace the missing values

In [14]:
# Identify duplicate rows
df[df.duplicated()]

Empty DataFrame
Columns: [Customer, State, Customer Lifetime Value, Response, Coverage, Education, Effective To Date, EmploymentStatus, Gender, Income, Location Code, Marital Status, Monthly Premium Auto, Months Since Last Claim, Months Since Policy Inception, Number of Open Complaints, Number of Policies, Policy Type, Policy, Renew Offer Type, Sales Channel, Total Claim Amount, Vehicle Class, Vehicle Size]
Index: []

[0 rows x 24 columns]

There are no duplicate rows, therefore no action is required to drop the duplicate rows or to determine if the duplicate like rows are actually consist of unique values

No changes have been made to the original form of the file